attestation_statement_apple.go

```diff
@@ -0,0 +1,76 @@
+package webauthn
+
+import (
+	"crypto"
+	"crypto/sha256"
+	"crypto/subtle"
+	"fmt"
+
+	"github.com/pomerium/webauthn/cose"
+)
+
+// VerifyAppleAttestationStatement verifies that an AttestationObject's attestation statement is valid according to the
+// packed verification procedure.
+func VerifyAppleAttestationStatement(
+	attestationObject *AttestationObject,
+	clientDataJSONHash ClientDataJSONHash,
+) error {
+	certificate, err := attestationObject.Statement.UnmarshalCertificate()
+	if err != nil {
+		return fmt.Errorf("%w: %s", ErrInvalidAttestationStatement, err)
+	}
+
+	authenticatorData, err := attestationObject.UnmarshalAuthenticatorData()
+	if err != nil {
+		return fmt.Errorf("%w: %s", ErrInvalidAttestationStatement, err)
+	}
+
+	publicKey, _, err := cose.UnmarshalPublicKey(
+		authenticatorData.AttestedCredentialData.CredentialPublicKey,
+	)
+	if err != nil {
+		return fmt.Errorf("%w: %s", ErrInvalidAttestationStatement, err)
+	}
+
+	// 2. Concatenate authenticatorData and clientDataHash to form nonceToHash.
+	nonceToHash := concat(attestationObject.AuthData, clientDataJSONHash[:])
+
+	// 3. Perform SHA-256 hash of nonceToHash to produce nonce.
+	nonce := sha256.Sum256(nonceToHash)
+
+	// 4. Verify that nonce equals the value of the extension with OID 1.2.840.113635.100.8.2 in credCert.
+	certificateNonce, err := getCertificateAppleNonce(certificate)
+	if err != nil {
+		return fmt.Errorf("%w: %s", ErrInvalidAttestationStatement, err)
+	}
+	if subtle.ConstantTimeCompare(nonce[:], certificateNonce) != 1 {
+		return fmt.Errorf("%w: invalid nonce", ErrInvalidAttestationStatement)
+	}
+
+	// 5. Verify that the credential public key equals the Subject Public Key of credCert.
+	err = verifyAppleAttestationStatementCredentialPublicKey(
```

Comment:
nit: i think we can return here directly

Reply:
I prefer not to do that since its inconsistent with what comes before. But sure I can change it.